### PR TITLE
Enhancement: optical jamming

### PR DESCRIPTION
**Pull request by Hurleveur**
**Feature:** This PR implements the feature discussed with Grey, optical jamming

## Feature Details
Optical jamming works in exactly the same way as radar jamming does.
Goal is to make jamming possible so they stop being a meme. One can adjust their heat and radar jamming but not optical yet, which this will allow.

## UI Screenshots
n/a

## Usage Examples
Going to be used for a hai optical jammer that works based on holograms, since the hai use trackers so much and they rely solely on optical tracking. They'll constantly use energy to function though.

## Testing Done
n/a

### Automated Tests Added
n/a

## Performance Impact
n/a

### CODE REVIEW

File: source/Projectile.cpp [hunk 1]
```diff
@@ -219,14 +219,21 @@ void Projectile::Move(vector<Visual> &visuals, vector<Projectile> &projectiles)
...
 		// to the target (jamming power attenuates with distance).
+		if(weapon->OpticalTracking())
+		{
+			double opticalTracking = weapon->OpticalTracking();
+			double opticalJamming = target->Attributes().Get("optical jamming");
+			if(!opticalJamming)
+				opticalConfused = Random::Real() > opticalTracking;
+			else
+				opticalConfused = Random::Real() > (opticalTracking * position.Distance(target->Position()))
+					/ (sqrt(opticalJamming) * weapon->Range());
```
Comment (quyykk) [inline source/Projectile.cpp:234]: Seems like this is the same as below, why not make move this into a separate function?

Comment (Hurleveur) [reply source/Projectile.cpp]: done

**Comment by Zitchas** [comment]
I always figured that the way we *should* be able to "jam" optical tracking is switching to a smaller ship. Fly something like a fighter, and optical tracking would be almost completely ineffective, while flying something like a heavy warship, and it's basically guaranteed to never miss.

I don't know if optical tracking *actually* works like that, (it certainly doesn't seem to) but I see this as erasing one of the few potential ways that small ships could actually have an advantage in combat.

That being said, optical jamming is good to have. I just don't want to see it become commonplace, as the main way to deal with optical tracking *should* be to get a smaller ship.

**Comment by warp-core** [comment]
> I always figured that the way we _should_ be able to "jam" optical tracking is switching to a smaller ship. Fly something like a fighter, and optical tracking would be almost completely ineffective, while flying something like a heavy warship, and it's basically guaranteed to never miss.
> 
> I don't know if optical tracking _actually_ works like that, (it certainly doesn't seem to) but I see this as erasing one of the few potential ways that small ships could actually have an advantage in combat.
> 
> That being said, optical jamming is good to have. I just don't want to see it become commonplace, as the main way to deal with optical tracking _should_ be to get a smaller ship.

That is how optical tracking works.

https://github.com/endless-sky/endless-sky/blob/51f22993b1753b01c0c843c7bac61e6fad32acc5/source/Projectile.cpp#L377-L383

**Comment by Hurleveur** [comment]
> That being said, optical jamming is good to have. I just don't want to see it become commonplace, as the main way to deal with optical tracking should be to get a smaller ship.

I agree with that. My idea is only the Hai will have it and as a very niche use, because it will constantly require power to operate, that said it's best than getting shot by a thousand trackers. Trying to use AM against that is just stupid, and since they have hologram tech I think it only makes sense they would develop a defence.

My idea of it is to cost like 100 power for a 8 outfit space jammer that would give as much jamming as a big jammer. The Unfettered version would cost 150 power for 30% more jamming or smth.

**Comment by vitalchip** [comment]
I think this is fine issofar as it adds an extra jamming type and it's done in keeping with the current mechanics.

Missiles need a rework, everyone knows it and there's no concensus on a way forward but I think this is an addition which would survive in some form regardless.

**Comment by Zitchas** [comment]
Just for interest sake, here's a graph of the optical tracking formula:

https://www.desmos.com/calculator/2rnqgoa6a4

While the comment in the code says that interceptors have a 15% chance of being tracked, it's worth noting that's for the Sparrow. The Hawk (using the Hai Tracker's value of 0.8) has closer to a 35% chance of being tracked.

File: source/Projectile.cpp [hunk 2]
```diff
@@ -219,23 +228,24 @@ void Projectile::Move(vector<Visual> &visuals, vector<Projectile> &projectiles)
...
-		// Optical: proportional tracking quality.
-		if(weapon->OpticalTracking())
-			opticalConfused = Random::Real() > weapon->OpticalTracking();
-
-		// Radar: If the target has no jamming, then proportional to tracking
+		// Optical and Radar: If the target has no jamming, then proportional to tracking
 		// quality. If the target does have jamming, then it's proportional to
 		// tracking quality, the strength of target's jamming, and the distance
 		// to the target (jamming power attenuates with distance).
+		if(weapon->OpticalTracking())
```
Comment (Amazinite) [inline source/Projectile.cpp:235]:
```suggestion
		double distance = position.Distance(target->Position());
		if(weapon->OpticalTracking())
```

File: source/Projectile.cpp [hunk 3]
```diff
@@ -219,23 +228,24 @@ void Projectile::Move(vector<Visual> &visuals, vector<Projectile> &projectiles)
...
+		{
+			double opticalTracking = weapon->OpticalTracking();
+			double opticalJamming = target->Attributes().Get("optical jamming");
+			opticalConfused = ConfusedTracking(opticalTracking, weapon->Range(),
+				opticalJamming, position.Distance(target->Position()));
```
Comment (Amazinite) [inline source/Projectile.cpp:240]:
```suggestion
				opticalJamming, distance);
```

File: source/Projectile.cpp [hunk 4]
```diff
@@ -219,23 +228,24 @@ void Projectile::Move(vector<Visual> &visuals, vector<Projectile> &projectiles)
...
 		{
 			double radarTracking = weapon->RadarTracking();
 			double radarJamming = target->Attributes().Get("radar jamming");
-			if(!radarJamming)
-				radarConfused = Random::Real() > radarTracking;
-			else
-				radarConfused = Random::Real() > (radarTracking * position.Distance(target->Position()))
-					/ (sqrt(radarJamming) * weapon->Range());
+			radarConfused = ConfusedTracking(radarTracking, weapon->Range(),
+				radarJamming, position.Distance(target->Position()));
```
Comment (Amazinite) [inline source/Projectile.cpp:248]:
```suggestion
				radarJamming, distance);
```

**Comment by warp-core** [comment]
Wait, was the other PR failing the style check because of line length?

**Pull request merged**